client/src/main/java/io/pravega/client/segment/impl/SegmentMetadataClientImpl.java

```diff
@@ -0,0 +1,265 @@
+            try {
+                c.getNow(null).close();
+            } catch (Exception e) {
+                log.warn("Exception tearing down connection: ", e);
+            }
+        }
+        failAllInflight(exceptionToInflightRequests);
+    }
+    
+    private void failAllInflight(Exception e) {
+        log.info("SegmentMetadata connection failed due to a {}.", e.toString());
+        List<CompletableFuture<StreamSegmentInfo>> infoRequestsToFail;
+        List<CompletableFuture<WireCommands.SegmentAttribute>> getAttributeRequestsToFail;
+        List<CompletableFuture<SegmentAttributeUpdated>> setAttributeRequestsToFail;
+        synchronized (lock) {
+            infoRequestsToFail = new ArrayList<>(infoRequests.values());
+            getAttributeRequestsToFail = new ArrayList<>(getAttributeRequests.values());
+            setAttributeRequestsToFail = new ArrayList<>(setAttributeRequests.values());
+            infoRequests.clear();
+            getAttributeRequests.clear();
+            setAttributeRequests.clear();
+        }
+        for (CompletableFuture<StreamSegmentInfo> infoRequest : infoRequestsToFail) {
+            infoRequest.completeExceptionally(e);
+        }
+        for (CompletableFuture<WireCommands.SegmentAttribute> getAttributeRequest : getAttributeRequestsToFail) {
+            getAttributeRequest.completeExceptionally(e);
+        }
+        for (CompletableFuture<SegmentAttributeUpdated> setAttributeRequest : setAttributeRequestsToFail) {
+            setAttributeRequest.completeExceptionally(e);
+        }
+    }
+    
+    CompletableFuture<ClientConnection> getConnection() {
+        synchronized (lock) {
+            //Optimistic check
+            if (connection != null) {
+                return connection;
+            }
+        }
+        return controller.getEndpointForSegment(segmentId.getScopedName()).thenCompose((PravegaNodeUri uri) -> {
+            log.info("Connecting to {}", uri);
+            synchronized (lock) {
+                if (connection == null) {
+                    connection = connectionFactory.establishConnection(uri, responseProcessor);
+                }
+                return connection; 
+            } 
+        });
+    }
+    
+    private CompletableFuture<WireCommands.StreamSegmentInfo> getSegmentInfo() {
+        CompletableFuture<WireCommands.StreamSegmentInfo> result = new CompletableFuture<>();
+        long requestId = requestIdGenerator.get();
+        synchronized (lock) {
+            infoRequests.put(requestId, result);
+        }
+        getConnection().thenAccept(c -> {
+            try {
+                log.debug("Getting segment info for segment: {}", segmentId);
+                c.send(new WireCommands.GetStreamSegmentInfo(requestId, segmentId.getScopedName()));
+            } catch (ConnectionFailedException e) {
+                closeConnection(e);
+            }
+        }).exceptionally(e -> {
```

Comment:
You can use `FutureHelpers.exceptionListener` to simplify this syntax (i.e., not having to add `return null` at the end).